aws/resource_aws_network_interface_sg_attachment_test.go

```diff
@@ -0,0 +1,238 @@
+package aws
+
+import (
+	"fmt"
+	"strconv"
+	"testing"
+
+	"github.com/hashicorp/terraform/helper/resource"
+	"github.com/hashicorp/terraform/terraform"
+)
+
+func TestAccAwsNetworkInterfaceSGAttachment(t *testing.T) {
+	cases := []struct {
+		Name                      string
+		CheckPrimaryInterfaceAttr bool
+		Config                    func(bool) string
+	}{
+		{
+			Name: "instance primary interface",
+			CheckPrimaryInterfaceAttr: false,
+			Config: testAccAwsNetworkInterfaceSGAttachmentConfigViaInstance,
+		},
+		{
+			Name: "externally supplied instance through data source",
+			CheckPrimaryInterfaceAttr: true,
+			Config: testAccAwsNetworkInterfaceSGAttachmentConfigViaDataSource,
+		},
+	}
+	for _, tc := range cases {
+		t.Run(tc.Name, func(t *testing.T) {
+			resource.Test(t, resource.TestCase{
+				PreCheck:  func() { testAccPreCheck(t) },
+				Providers: testAccProviders,
+				Steps: []resource.TestStep{
+					resource.TestStep{
+						Config: tc.Config(true),
+						Check:  checkSecurityGroupAttachment(tc.CheckPrimaryInterfaceAttr, true),
+					},
+					resource.TestStep{
+						Config: tc.Config(false),
+						Check:  checkSecurityGroupAttachment(tc.CheckPrimaryInterfaceAttr, false),
+					},
+				},
+			})
+		})
+	}
+}
+
+func checkSecurityGroupAttachment(checkPrimaryInterfaceAttr bool, expected bool) resource.TestCheckFunc {
```

Comment:
Probably a nitpick, but where possible string arguments are usually better in readability than boolean ones as it's more obvious how do they affect the behaviour without having to read the body of the function (related to `checkPrimaryInterfaceAttr`).